staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go

```diff
@@ -0,0 +1,206 @@
+		requestContextMapper: requestContextMapper,
+		sink:                 sink,
+		policy:               policy,
+	}
+	// check if the ResponseWriter we're wrapping is the fancy one we need
+	// or if the basic is sufficient
+	_, cn := responseWriter.(http.CloseNotifier)
+	_, fl := responseWriter.(http.Flusher)
+	_, hj := responseWriter.(http.Hijacker)
+	if cn && fl && hj {
+		return &fancyFailedAuthnResponseWriterDelegator{delegate}
+	}
+	return delegate
+}
+
+var _ http.ResponseWriter = &auditFailedAuthnResponseWriter{}
+
+// auditFailedAuthnResponseWriter intercepts WriteHeader, sets it in the event.
+type auditFailedAuthnResponseWriter struct {
+	http.ResponseWriter
+	event                *auditinternal.Event
+	once                 sync.Once
+	username             string
+	request              *http.Request
+	requestContextMapper request.RequestContextMapper
+	sink                 audit.Sink
+	policy               policy.Checker
+}
+
+func (a *auditFailedAuthnResponseWriter) setHttpHeader() {
+	if a.event != nil {
+		a.ResponseWriter.Header().Set(auditinternal.HeaderAuditID, string(a.event.AuditID))
+	}
+}
+
+func (a *auditFailedAuthnResponseWriter) Write(bs []byte) (int, error) {
+	// the Go library calls WriteHeader internally if no code was written yet. But this will go unnoticed for us
+	a.processCode(http.StatusOK)
+	a.setHttpHeader()
+	return a.ResponseWriter.Write(bs)
+}
+
+func (a *auditFailedAuthnResponseWriter) WriteHeader(code int) {
+	a.processCode(code)
+	a.setHttpHeader()
+	a.ResponseWriter.WriteHeader(code)
+}
+
+func (a *auditFailedAuthnResponseWriter) processCode(code int) {
+	a.once.Do(func() {
+		ctx, ok := a.requestContextMapper.Get(a.request)
+		if !ok {
+			responsewriters.InternalError(a.ResponseWriter, a.request, errors.New("no context found for request"))
+			return
+		}
+
+		// if there already exists an audit event in the context we don't need to do anything
+		if ae := request.AuditEventFrom(ctx); ae != nil {
+			return
+		}
+
+		// otherwise, we need to create the event and log the authn error
+		ev, err := createAuditEvent(ctx, a.policy, a.requestContextMapper, a.request, a.ResponseWriter)
+		if err != nil {
+			return
+		}
+		if ev == nil {
+			return
+		}
+
+		// since user is not set at this point, we need to read it manually
+		ev.User.Username = a.username
+		ctx = request.WithAuditEvent(ctx, ev)
+		if err := a.requestContextMapper.Update(a.request, ctx); err != nil {
+			utilruntime.HandleError(fmt.Errorf("failed to attach audit event to the context: %v", err))
+			responsewriters.InternalError(a.ResponseWriter, a.request, errors.New("failed to update context"))
+			return
+		}
+
+		ev.ResponseStatus = &metav1.Status{}
+		ev.ResponseStatus.Code = int32(code)
+		ev.Stage = auditinternal.StageResponseStarted
+		processAuditEvent(a.sink, ev)
+		a.event = ev
+	})
+}
+
+// getUsername returns username or information on the authn method being used.
+func getUsername(req *http.Request) string {
```

Comment:
AIUI we get an authentication failure, then handover the failed request to this code to try and extract some context from the request. Is there something stopping us logging a more meaningful failure at the actual authentication failure time? Seems like the point where we determine the password/cert/etc is bad is the place where we should create the log.

This hard-codes some assumptions about the auth methods we accept which would seem to better belong in a single library (or libraries, one for each method) so that we can be careful about how we handle this untrusted input and reason about code changes needed to deprecate authentication methods if necessary.

Reply:
Hrm... I'm not a big fan of it either, but it was the first thing I could think of. I'll dig into the code once more and see how we can do a better job here.